File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/3273-두 수의 합/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
typedef long long int ll;
using namespace std;

#define NM 100005

int N, Sum;
int A[NM];

void input() {
	cin >> N;
	for (int i = 1; i <= N; i++) cin >> A[i];
	cin >> Sum;
}

int bin_search(int A[], int L, int R, int X) {
	while (L <= R) {
		int mid = (L + R) / 2;
		if (A[mid] == X)
			return true;

		if (A[mid] < X)
			L = mid + 1;
		else
			R = mid - 1;
	}
	return false;
}

void pro() {
	sort(A + 1, A + 1 + N);
	int ans = 0;
	for (int i = 1; i <= N - 1; i++) {
		if (bin_search(A, i + 1, N, Sum - A[i])) ans++;
	}
	cout << ans;
}


int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	pro();
	return 0;
}

File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/3273-두 수의 합/solution.py
import sys
si = sys.stdin.readline
n = int(si())
a = list(map(int, si().split()))
sum = int(si())

def bin_search(a, l, r, x):
    while l <= r:
        mid = (l + r) // 2
        if a[mid] == x:
            return True
        if a[mid] < x: l = mid + 1
        else: r = mid - 1
    return False

a.sort()
ans = 0
for i in range(n - 1):
    if bin_search(a, i + 1, n - 1, sum - a[i]): ans += 1
    
print(ans)